### ISSUE TITLE
Syntax error with type parameter and array literal

### ISSUE THREAD

**Issue by hexane360** [JuliaLang/julia#38681]
I recently ran into a weird parsing error when defining the following function:
```
julia> function to_array(val::T)::Array{T} where {T} [val] end
ERROR: syntax: space before "[" not allowed in "{T} [" at REPL[14]:1
Stacktrace:
 [1] top-level scope at none:1
```
If I add a newline before the function body, it parses and runs fine.

If I remove the type parameter, I get the same syntax error but with the return type:
```
julia> function to_arr(val::T)::Array{T} [val] end
ERROR: syntax: space before "[" not allowed in "Array{T} [" at REPL[27]:1
Stacktrace:
 [1] top-level scope at none:1
```
Is this expected behavior? I can't think of any language feature that allows square brackets after type parameters.

Version info:
```
julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: AMD Ryzen 7 2700X Eight-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, znver1)
Environment:
  JULIA_EDITOR = emacsclient.exe
```

**Comment by FedericoStra** [comment]
The second example `function to_arr(val::T)::Array{T} [val] end` is definitely wrong (even formatting it on multiple lines), because it is missing the quantifier [`where`](https://docs.julialang.org/en/v1/base/base/#Keywords) required for [parametric methods](https://docs.julialang.org/en/v1/manual/methods/).

---

> I can't think of any language feature that allows square brackets after type parameters.

```
Union{Int,Char}[1, 2, 'a', 42, 'x']
```

**Comment by hexane360** [comment]
>The second example function to_arr(val::T)::Array{T} [val] end is definitely wrong (even formatting it on multiple lines), because it is missing the quantifier where required for parametric methods.

Yes, but this shouldn't result in a *syntax* error; the error should be an `UndefVarError` later on.

>Typed array literals:

Ah, I had forgotten about those.

I don't think this is purely a precedence issue, because the same thing happens with parentheses around the array literal:
```
julia> function to_array(val::T)::Array{T} where {T} ([val]) end
ERROR: syntax: space before "(" not allowed in "{T} (" at REPL[14]:1
Stacktrace:
 [1] top-level scope at none:1
```

It seems like sort of a "garden-path" situation, where the parser sees an opening bracket (or parentheses) and immediately decides to parse it as a continuation of the type, rather than as the start of the block. Then it fails to do so and promptly returns a syntax error.

Interestingly, any constructions without the space return a deprecation syntax error:
```
julia> {T}[5]
ERROR: syntax: { } vector syntax is discontinued around REPL[30]:1
Stacktrace:
 [1] top-level scope at REPL[30]:1
julia> {T}(5)
ERROR: syntax: { } vector syntax is discontinued around REPL[31]:1
Stacktrace:
 [1] top-level scope at REPL[31]:1
julia> {T}{5}
ERROR: syntax: { } vector syntax is discontinued around REPL[31]:1
Stacktrace:
 [1] top-level scope at REPL[32]:1
```
